zap/src/main/java/org/parosproxy/paros/core/scanner/HostProcess.java

```diff
@@ -313,12 +314,23 @@ public void run() {
             }
 
             for (StructuralNode startNode : startNodes) {
+                Map<String, Integer> indexOfTemporary = new HashMap<>();
                 traverse(
                         startNode,
                         true,
                         node -> {
                             if (canScanNode(node)) {
-                                messagesIdsToAppScan.add(node.getHistoryReference().getHistoryId());
+                                Integer index = indexOfTemporary.get(nodeHash(node));
+                                int nodeHistoryId = node.getHistoryReference().getHistoryId();
+                                if (isTemporary(node) && index == null) {
+                                    indexOfTemporary.put(
+                                            nodeHash(node), messagesIdsToAppScan.size());
+                                    messagesIdsToAppScan.add(nodeHistoryId);
+                                } else if (!isTemporary(node) && index == null) {
+                                    messagesIdsToAppScan.add(nodeHistoryId);
+                                } else if (!isTemporary(node) && index != null) {
+                                    messagesIdsToAppScan.set(index, nodeHistoryId);
+                                }
```

Comment:
The checks can be reorder instead, e.g.:
```Java
if (index == null) {
    if (isTemporary(node)) {
        indexOfTemporary.put(nodeHash(node), messagesIdsToAppScan.size());
    }
    messagesIdsToAppScan.add(nodeHistoryId);
} else if (!isTemporary(node)) {
    messagesIdsToAppScan.set(index, nodeHistoryId);
}
```

Reply:
yes, that looks better.